### ISSUE TITLE
RC2: Feature Interaction & Information Decay Analysis (6A §3b)

### ISSUE THREAD

**Issue by Dmytro-Khvedchuk** [Dmytro-Khvedchuk/RSPCP_bachelors_thesis#61]
## RC2: Feature Interaction & Information Decay Analysis (6A §3b)

> Insert after §3 in `research/RC2_features_and_profiling.ipynb`
> **New analysis** — goes beyond marginal MI to joint information and horizon dynamics. Be as thorough as needed.

---

### Theoretical Context

**Conditional Mutual Information (Cover & Thomas, 2006):** Marginal MI measures `I(X_j; Y)` — how much *one* feature tells you about the target. But features interact. Two features that individually carry zero MI can jointly carry significant MI if they encode complementary aspects of the signal (e.g., momentum direction + volatility magnitude). Conditional MI: `I(X_j; Y | X_k) = I(X_j, X_k; Y) - I(X_k; Y)` measures the *additional* information feature j provides given feature k is already known. If `I(X_j; Y | X_k) >> I(X_j; Y)`, features j and k are synergistic. If `I(X_j; Y | X_k) ≈ 0`, feature j is redundant given k.

**Information Decay (Lillo & Farmer, 2004; Cont, 2005):** In financial markets, predictive information decays with horizon — today's momentum signal tells you something about tomorrow's return but nothing about next month's. The *decay curve* (MI or DA as a function of forecast horizon) reveals the "sweet spot" where signal peaks. For crypto dollar bars, this is likely 1-4 bars (hours to a day). Beyond that, noise dominates. Plotting the decay curve informs the critical horizon selection decision for Phase 9-10.

**Feature Clustering (Tibshirani et al., 2001; Hastie et al., 2009):** Features that share MI with the target in similar patterns form *information groups*. Hierarchical clustering on MI-based distances reveals these groups. If the recommender (Phase 12) uses 5 momentum features that are all in the same cluster, it's wasting capacity. One representative per cluster is sufficient. This directly supports the VIF analysis from §3 — VIF catches linear redundancy, MI clustering catches nonlinear redundancy.

---

#### 3b.1: Conditional MI Matrix (~10 cells)

For the **top 10 features** by marginal MI:
- Compute `I(X_j; Y | X_k)` for all 10×10 pairs
- **Synergy heatmap:** color = `I(X_j; Y | X_k) - I(X_j; Y)` (positive = synergy, negative = redundancy)
- Identify the most synergistic pair: "Combining [feature_a] and [feature_b] provides X% more information than either alone"
- **Therefore:** "Features [A, B] are synergistic — both should be retained even if individually weak. Features [C, D] are redundant — keeping both adds no information."

#### 3b.2: Information Decay Curves (~10 cells)

For the **top 5 features** by MI, across horizons `{1, 2, 4, 8, 12, 24}` bars:
- **MI decay plot:** x = horizon (log scale), y = MI (nats), one line per feature
- **DA decay plot:** x = horizon, y = DA excess over 50%, one line per feature
- Identify the "sweet spot horizon" where signal peaks
- Identify features that decay fast (short-memory) vs slow (long-memory)
- **Therefore:** "Signal peaks at horizon [H] bars. Features [X, Y] retain signal to horizon 24 (long-memory indicators like Hurst). Features [Z, W] decay by horizon 4 (short-memory momentum). This informs multi-horizon modeling in Phase 9-10."

#### 3b.3: Feature Clustering Dendrogram (~5 cells)

- Distance metric: `d(i,j) = 1 - I(X_i; X_j) / max(H(X_i), H(X_j))` (normalized MI distance)
- **Dendrogram:** hierarchical clustering (Ward linkage) showing feature groups
- Cut at threshold → identify N clusters → one representative per cluster
- Compare with VIF clusters: are the same features flagged by both methods?
- **Therefore:** "Features form N information groups. Selecting one representative per group yields [M] features — this is the effective feature dimensionality."

---

### Chart & Table Catalog

| # | Output | Type | Axes | Interpretation |
|---|--------|------|------|----------------|
| 1 | Conditional MI synergy heatmap | `sns.heatmap` | top-10 features × top-10 features, color = synergy | Red = synergistic, blue = redundant |
| 2 | Top synergistic pairs table | DataFrame | feature_a, feature_b, MI_joint, MI_a, MI_b, synergy_ratio | Which pairs to definitely keep together |
| 3 | MI decay curves | Line plot | x = horizon (log), y = MI, colored by feature | "Sweet spot" horizon visible as peak |
| 4 | DA decay curves | Line plot | x = horizon, y = DA_excess_pp, colored by feature | Economically interpretable version |
| 5 | Feature dendrogram | `scipy.cluster.hierarchy.dendrogram` | Hierarchical tree | Information groups at a glance |
| 6 | Cluster membership table | DataFrame | cluster_id, members, representative, MI_representative | One rep per cluster for dimensionality reduction |

---

### Code Hints

```python
# Conditional MI: sklearn.feature_selection.mutual_info_regression
#   with residualization or direct multivariate estimation
# Simpler alternative: fit MI(X_j; Y) and MI(X_j; Y | X_k) via
#   binning + entropy estimation (npeet package or custom KNN estimator)
# Information decay: loop over horizons, call MIPermutationValidator
#   from src/app/features/application/validation.py with target_col override
# Feature clustering: scipy.cluster.hierarchy (Ward linkage)
# Normalized MI distance: sklearn.metrics.normalized_mutual_info_score
```

---

### Acceptance Criteria

- [ ] Conditional MI matrix computed for top 10 features (10×10)
- [ ] Synergy heatmap rendered with clear color legend (red = synergy, blue = redundancy)
- [ ] At least one synergistic pair identified and named
- [ ] MI decay curves plotted for top 5 features across horizons {1, 2, 4, 8, 12, 24}
- [ ] DA decay curves show the same horizons with economically interpretable y-axis
- [ ] "Sweet spot" horizon explicitly identified in narrative
- [ ] Feature dendrogram rendered with clear cluster cut threshold
- [ ] Cluster membership table lists representatives per cluster
- [ ] VIF vs MI clustering comparison noted (do they agree on redundant features?)
- [ ] All three "Therefore:..." paragraphs present
- [ ] All outputs render correctly and tell a coherent story

---

### Examiner-Proofing

- *"Did you check for feature interactions?"* → Conditional MI matrix with synergy/redundancy heatmap
- *"How did you choose your forecast horizon?"* → Information decay curves showing signal peak
- *"Are your features really independent, or are they measuring the same thing?"* → Feature clustering dendrogram with effective dimensionality count
- *"You have 23 features — isn't that too many for 5000 samples?"* → Effective dimensionality from clustering: "23 features collapse into N information groups"

### Negative Result Protocol

- If **no synergistic pairs found**: "Features carry independent marginal information — no interaction effects detected. Feature selection can proceed on marginal MI alone."
- If **MI decays to zero by horizon 2**: "Signal is extremely short-lived. Only horizon-1 modeling is viable. Multi-horizon approach (4, 24) may not be supported."
- If **all features in one cluster**: Severe redundancy. Consider PCA or selecting only the highest-MI representative.

### Cross-References

- **Upstream:** #45 (§3 MI results, VIF analysis), Phase 4D (#36 — validation results)
- **Downstream:** #48 (Go/No-Go — horizon selection, effective dimensionality), Phase 9-10 (horizon choice, feature subset)
- **Literature:** Cover & Thomas (2006), Lillo & Farmer (2004), Cont (2005), Tibshirani et al. (2001)
